### ISSUE TITLE
Rule proposal: Disallow implicit spaces between JSX strings and anything else

### ISSUE THREAD

**Issue by deniseli** [yannickcr/eslint-plugin-react#918]
In the interest of being defensive against refactor-induced bugs, could we disallow spaces between JSX strings and everything else? They're very hard for people to notice. For example:
```
render() {
  return (
    <span>
      Welcome, <Name userid='blahblah' />
    </span>
  );
}
```
While refactoring, someone could easily make the mistake of making the code look like this:
```
render() {
  return (
    <span>
      Welcome,
      <Name userid='blahblah' />
    </span>
  );
}
```
This removes the space after "Welcome," which messes up the spacing.
The style I've been enforcing (within my organization) is to instead do
```
render() {
  return (
    <span>
      Welcome,
      {' '}
      <Name userid='blahblah' />
    </span>
  );
}
```
While uglier, this does make it very clear that the space belongs there, and prevents it from being accidentally removed.

**Comment by lencioni** [comment]
Just to be clear, this proposal is to flag the following as an error:

```jsx
render() {
  return (
    <span>
      Welcome, <Name userid='blahblah' />
    </span>
  );
}
```

in favor of this:

```jsx
render() {
  return (
    <span>
      Welcome,
      {' '}
      <Name userid='blahblah' />
    </span>
  );
}
```

Is that accurate?

**Comment by ljharb** [comment]
I think it should also allow `Welcome,{' '}<Name userid='blahblah' />` though, no?

**Comment by deniseli** [comment]
Yes, and also `{'Welcome, '}<...>`. The only critical part is disallowing the space when it isn't wrapped.

**Comment by ljharb** [comment]
This seems similar to (but different from) [jsx-no-literals](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-literals.md).

**Comment by deniseli** [comment]
@ljharb Yeah. That rule would also fix the issue. It's just a stricter rule than is necessary to fix this particular issue.
